### PR TITLE
persist energyUsage and fix some issue.

### PR DESCRIPTION
**Pull request by liizfq**
## Description

Please carefully read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md) before making any pull requests.
And, **Do not make a pull request to merge into stable unless it is a hotfix. Use the development branch instead.**
## Type of changes

- [ ] Bug fix
- [ ] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] My pull request is unique and no other pull requests have been opened for these changes
- [ ] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [ ] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

File: src/main/java/emu/grasscutter/game/avatar/Avatar.java [hunk 1]
```diff
@@ -359,30 +360,34 @@ public void setCurrentHp(float currentHp) {
...
 		}		
 	}
 
 	public void setCurrentEnergy(FightProperty curEnergyProp, float currentEnergy) {
-		if (GAME_OPTIONS.energyUsage) {
-			this.setFightProperty(curEnergyProp, currentEnergy);
-			this.currentEnergy = currentEnergy;
+		this.setFightProperty(curEnergyProp, currentEnergy);
+		this.currentEnergy = currentEnergy;
+		if (this.getPlayer().getEnergyUsage()) {
```
Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/game/avatar/Avatar.java]: This sets the energy irrespective of whether energy usage is turned on or off. That seems like a bug?

File: src/main/java/emu/grasscutter/game/entity/EntityAvatar.java [hunk 2]
```diff
@@ -175,12 +175,28 @@ public void addEnergy(float amount, PropChangeReason reason, boolean isFlat) {
...
 		}
-	} 
+		setCurrentEnergy(curEnergyProp,curEnergy,reason);
+	}
+
+	private void setCurrentEnergy(FightProperty curEnergyProp,float newEnergy,PropChangeReason reason) {
+		this.avatar.setCurrentEnergy(curEnergyProp, newEnergy);
+		this.getScene().broadcastPacket(new PacketAvatarFightPropUpdateNotify(this.getAvatar(), curEnergyProp));
+		this.getScene().broadcastPacket(new PacketEntityFightPropChangeReasonNotify(this, curEnergyProp, newEnergy, reason));
+	}
```
Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/game/entity/EntityAvatar.java]: It is confusing that there is now a `setCurrentEnergy` for both `Avatar` and `EntityAvatar`. Please unify them and put them into either one of those classes, probably in `EntityAvatar`.

File: src/main/java/emu/grasscutter/game/managers/energy/EnergyManager.java [hunk 3]
```diff
@@ -152,12 +187,19 @@ public void handleGenerateElemBall(AbilityInvokeEntry invoke) throws InvalidProt
...
         // Generate 2 particles by default.
-        int amount = 2;
+        int amount = particlesDefault;
+        if(invoke.getHead().getLocalId()==98307){
+            amount = 3;
+            for (int i = 0; i < amount; i++) {
+                generateElemBall(itemId, new Position(action.getPos()), 1);
+            }
+            return;
+        }
```
Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/game/managers/energy/EnergyManager.java:165]: What is that? A fix for Favonius? Please put a discriptive comment if you handle a specific special case.

File: src/main/java/emu/grasscutter/utils/ConfigContainer.java [hunk 4]
```diff
@@ -176,7 +176,7 @@ public static class GameOptions {
         public boolean watchGachaConfig = false;
         public boolean enableShopItems = true;
         public boolean staminaUsage = true;
-        public boolean energyUsage = false;
+//        public boolean energyUsage = false;
```
Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/utils/ConfigContainer.java]: I don't get what the point or benefit of this change is supposed to be. On the contrary, I can only see this as a detriment. What does taking away server manager's ability to set a default for whether they want to use such a - highly experimental, by the way - feature improve?

Giving the ability to override this config with a command is fine. Persisting that is fine too (though you would have to devise a way to differentiate between nothing being set and the feature being turned off). But taking away the configurability is not fine.

File: src/main/java/emu/grasscutter/game/avatar/Avatar.java [hunk 5]
```diff
@@ -359,30 +360,34 @@ public void setCurrentHp(float currentHp) {
...
-		}
+	public float getCurrentEnergy() {
+		return currentEnergy;
 	}
+
+//	public void setCurrentEnergy() {
+//		if (this.getPlayer().getEnergyUsage()) {
+//			this.setCurrentEnergy(this.currentEnergy);
+//		}
+//	}
```
Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/game/avatar/Avatar.java]: Leftover.

File: src/main/java/emu/grasscutter/game/entity/EntityAvatar.java [hunk 6]
```diff
@@ -175,12 +175,28 @@ public void addEnergy(float amount, PropChangeReason reason, boolean isFlat) {
...
+		FightProperty curEnergyProp = this.getAvatar().getSkillDepot().getElementType().getCurEnergyProp();
+		FightProperty maxEnergyProp = this.getAvatar().getSkillDepot().getElementType().getMaxEnergyProp();
+		PropChangeReason reason=PropChangeReason.PROP_CHANGE_REASON_ABILITY;
+		if(this.getPlayer().getEnergyUsage()){
+			curEnergy=this.avatar.getCurrentEnergy();
+		}else{
+			curEnergy= this.getFightProperty(maxEnergyProp);
 		}
-	} 
+		setCurrentEnergy(curEnergyProp,curEnergy,reason);
```
Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/game/entity/EntityAvatar.java]: This violates the protocol. On avatar creation, when this is called by the constructor, the live server sends neither an `AvatarFightPropUpdateNotify` nor an `EntityFightPropChangeReasonNotify`.

**Review by GanyusLeftHorn — CHANGES_REQUESTED** [review]

File: src/main/java/emu/grasscutter/game/entity/EntityAvatar.java [hunk 7]
```diff
@@ -175,12 +182,13 @@ public void addEnergy(float amount, PropChangeReason reason, boolean isFlat) {
 		
 		// Set energy and notify.
 		if (newEnergy != curEnergy) {
-			this.avatar.setCurrentEnergy(curEnergyProp, newEnergy);
-			
-			this.getScene().broadcastPacket(new PacketAvatarFightPropUpdateNotify(this.getAvatar(), curEnergyProp));
-			this.getScene().broadcastPacket(new PacketEntityFightPropChangeReasonNotify(this, curEnergyProp, newEnergy, reason));
+			this.avatar.setCurrentEnergy(newEnergy);
+			this.getScene().broadcastPacket(new PacketEntityFightPropUpdateNotify(this, curEnergyProp));
```
Comment (GanyusLeftHorn) [inline src/main/java/emu/grasscutter/game/entity/EntityAvatar.java]: Adding energy should send `AvatarFightPropUpdateNotify`.

Comment (liizfq) [reply src/main/java/emu/grasscutter/game/entity/EntityAvatar.java]: ok,but why send AvatarFightPropUpdateNotify before setCurrentEnergy?

Comment (GanyusLeftHorn) [reply src/main/java/emu/grasscutter/game/entity/EntityAvatar.java]: What do you mean?

Comment (GanyusLeftHorn) [reply src/main/java/emu/grasscutter/game/entity/EntityAvatar.java]: What do you mean?

Comment (liizfq) [reply src/main/java/emu/grasscutter/game/entity/EntityAvatar.java]: i guess that this.avatar.setCurrentEnergy(newEnergy)  executes before AvatarFightPropUpdateNotify.
AvatarFightPropUpdateNotify should notify the newEnergy.

Comment (GanyusLeftHorn) [reply src/main/java/emu/grasscutter/game/entity/EntityAvatar.java]: > i guess that this.avatar.setCurrentEnergy(newEnergy) executes before AvatarFightPropUpdateNotify.

Correct.

> AvatarFightPropUpdateNotify should notify the newEnergy.

Also correct.

I know it's slightly confusing, but there is a difference in how the official server handles communicating energy changes between adding and clearing energy. For adding energy, it sends `AvatarFightPropUpdateNotify`, and for clearing it sends `EntityFightPropUpdateNotify `

**Comment by 1-lavender-1** [comment]
Why is this PR closed？I consider these new features are quite useful

**Comment by liizfq** [comment]
> Why is this PR closed？I consider these new features are quite useful

i tried to fix conflict,but i failed.So i plan fork the lasted repository and add the new features.thanks for your reply.

**Pull request closed without merging**